File: Julia/textbook_track/chapter01/ba1d.jl
# Find All Occurrences of a Pattern in a String

function find_pattern_occurrences(text, pattern)
    k = length(pattern)
    matching_indices = Int64[]
    for ix in 1:(length(text)-k+1)
        substring = text[ix:ix+k-1]
        if substring == pattern
            push!(matching_indices, ix - 1)
        end
    end
    matching_indices
end


function main()
    pattern = readline()
    genome = readline()
    result = find_pattern_occurrences(genome, pattern)
    println(join(result, " "))
end


main()

File: OCaml/src/textbook_track/chapter01/ba1d.ml
let find_pattern_occurrences (text : string) (pattern : string) : int list =
    let text_length = String.length text in
    let pattern_length = String.length pattern in
    let rec loop acc ix =
        if ix = text_length - pattern_length + 1 then List.rev acc
        else
            let acc' = if (String.sub text ix pattern_length) = pattern then ix :: acc else acc in
            loop acc' (ix + 1) in
    loop [] 0


let () =
    let pattern = read_line () in
    let text = read_line () in
    let result = find_pattern_occurrences text pattern in
    result |> List.map string_of_int |> String.concat " " |> print_endline
